crypto/Pos/Crypto/Orphans.hs

```diff
@@ -8,51 +8,149 @@ import           Universum
+      fromByteString = BA.convert
+    in
+      contain $ fromByteString <$> safeGet
+
+
+instance SafeCopy BA.ScrubbedBytes where
+  putCopy s =
+    let
+      toByteString :: BA.ScrubbedBytes -> BS.ByteString
+      toByteString = BA.convert
+    in
+      contain $ safePut (toByteString s)
+  getCopy =
+    let
+      fromByteString :: BS.ByteString -> BA.ScrubbedBytes
+      fromByteString = BA.convert
+    in
+      contain $ fromByteString <$> safeGet
 
-instance NFData Ed25519.PublicKey
-instance NFData Ed25519.SecretKey
-instance NFData Ed25519.Signature
 
 deriveSafeCopySimple 0 'base ''Ed25519.PublicKey
 deriveSafeCopySimple 0 'base ''Ed25519.SecretKey
 deriveSafeCopySimple 0 'base ''Ed25519.Signature
 
+fromByteString :: BS.ByteString -> BA.Bytes
+fromByteString = BA.convert
+
 instance FromJSON Ed25519.PublicKey where
-    parseJSON v = Ed25519.PublicKey . getJsonByteString <$> parseJSON v
+  parseJSON v =
+    let
+      fromCryptoToAeson :: CryptoFailable a -> Parser a
+      fromCryptoToAeson (CryptoFailed e) = error $ mappend "Pos.Crypto.Orphan.parseJSON Ed25519.PublicKey failed because " (T.pack $ show e)
+      fromCryptoToAeson (CryptoPassed r) = return r
+    in
+      do
+        res <- Ed25519.publicKey . fromByteString . getJsonByteString <$> parseJSON v
+        fromCryptoToAeson res
 
 instance ToJSON Ed25519.PublicKey where
-    toJSON = toJSON . JsonByteString . Ed25519.openPublicKey
+  toJSON =
+    let
+      toByteString :: Ed25519.PublicKey -> BS.ByteString
+      toByteString = BA.convert
+    in
+      toJSON . JsonByteString . toByteString
 
 instance FromJSON Ed25519.Signature where
-    parseJSON v = Ed25519.Signature . getJsonByteString <$> parseJSON v
+  parseJSON v =
+    let
+      fromCryptoToAeson :: CryptoFailable a -> Parser a
+      fromCryptoToAeson (CryptoFailed e) = error $ mappend "Pos.Crypto.Orphan.parseJSON Ed25519.Signature failed because " (T.pack $ show e)
+      fromCryptoToAeson (CryptoPassed r) = return r
+    in
+      do
+        res <- Ed25519.signature . fromByteString . getJsonByteString <$> parseJSON v
+        fromCryptoToAeson res
 
 instance ToJSON Ed25519.Signature where
-    toJSON = toJSON . JsonByteString . Ed25519.unSignature
+  toJSON =
+    let
+      toByteString :: Ed25519.Signature -> BS.ByteString
+      toByteString = BA.convert
+    in
+      toJSON . JsonByteString . toByteString
+
+
+fromCryptoToDecoder :: T.Text -> CryptoFailable a -> Decoder s a
+fromCryptoToDecoder item (CryptoFailed e) = error $ "Pos.Crypto.Orphan.decode " <> item <> " failed because " <> show e
+fromCryptoToDecoder _ (CryptoPassed r) = return r
 
 instance Bi Ed25519.PublicKey where
-    encode (Ed25519.PublicKey k) = encode k
-    decode = Ed25519.PublicKey <$> decode
+    encode =  E.encodeBytes . BA.convert
+    decode =
+        do
+          res <- Ed25519.publicKey . fromByteString <$> decode
+          fromCryptoToDecoder "Ed25519.PublicKey" res
```

Comment:
style: to avoid needless indent, we tend to put a `do` on the `=` line, eg:

```haskell
instance ...
    decode = do
        res <- ...
        fromCryptoToDecoder ...
```